File: app/src/main/java/com/hover/stax/bounties/BountyEmailFragment.java
package com.hover.stax.bounties;

import static org.koin.java.KoinJavaComponent.get;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;

import com.hover.stax.R;
import com.hover.stax.databinding.FragmentBountyEmailBinding;
import com.hover.stax.navigation.NavigationInterface;
import com.hover.stax.utils.Utils;
import com.hover.stax.utils.network.NetworkMonitor;
import com.hover.stax.views.AbstractStatefulInput;
import com.hover.stax.views.StaxDialog;
import com.hover.stax.views.StaxTextInputLayout;

import java.lang.ref.WeakReference;
import java.util.Map;

public class BountyEmailFragment extends Fragment implements NavigationInterface, View.OnClickListener, BountyAsyncCaller.AsyncResponseListener {

    private static final String TAG = "BountyEmailFragment";
    private StaxTextInputLayout emailInput;
    private FragmentBountyEmailBinding binding;
    private StaxDialog dialog;

    private NetworkMonitor networkMonitor = get(NetworkMonitor.class);

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentBountyEmailBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        emailInput = binding.emailInput;
        emailInput.setText(Utils.getString(BountyActivity.EMAIL_KEY, requireActivity()));
        binding.continueEmailBountyButton.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        if (networkMonitor.isNetworkConnected()) {
            Utils.logAnalyticsEvent(getString(R.string.clicked_bounty_email_continue_btn), requireContext());
            if (validates()) {
                emailInput.setEnabled(false);
                new BountyAsyncCaller(new WeakReference<>(requireContext()), this).execute(emailInput.getText());
                emailInput.setState(getString(R.string.bounty_uploading_email), AbstractStatefulInput.INFO);
            } else {
                emailInput.setState(getString(R.string.bounty_email_error), AbstractStatefulInput.ERROR);
            }
        } else {
            showOfflineDialog();
        }
    }

    private void showOfflineDialog() {
        dialog = new StaxDialog(requireActivity())
                .setDialogTitle(R.string.internet_required)
                .setDialogMessage(R.string.internet_required_bounty_desc)
                .setPosButton(R.string.btn_ok, null)
                .makeSticky();
        dialog.showIt();
    }

    private void showEdgeCaseErrorDialog() {
        dialog = new StaxDialog(requireActivity())
                .setDialogMessage(getString(R.string.edge_case_bounty_email_error))
                .setPosButton(R.string.btn_ok, null);
        dialog.showIt();
    }

    private boolean validates() {
        if (emailInput.getText() == null) return false;
        String email = emailInput.getText().replace(" ", "");
        return email.matches("(?:[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)])");
    }

    @Override
    public void onComplete(Map<Integer, String> responseMap) {
        Map.Entry<Integer, String> entry = responseMap.entrySet().iterator().next();
        int responseCode = entry.getKey();
        String message = entry.getValue();
        if (responseCode >= 200 && responseCode < 300)
            saveAndContinue();
        else {
            Utils.logErrorAndReportToFirebase(TAG, message, null);

            if (isAdded()) {
                if (networkMonitor.isNetworkConnected()) showEdgeCaseErrorDialog();
                else setEmailError();
            }
        }
    }

    private void setEmailError() {
        Utils.logAnalyticsEvent(getString(R.string.bounty_email_err, getString(R.string.bounty_api_internet_error)), requireActivity());
        emailInput.setEnabled(true);
        emailInput.setState(getString(R.string.bounty_api_internet_error), AbstractStatefulInput.ERROR);
    }

    private void saveAndContinue() {
        Utils.logAnalyticsEvent(getString(R.string.bounty_email_success), requireContext());
        Utils.saveString(BountyActivity.EMAIL_KEY, emailInput.getText(), requireActivity());

        NavHostFragment.findNavController(this).navigate(R.id.bountyListFragment);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (dialog != null && dialog.isShowing())
            dialog.dismiss();

        binding = null;
    }
}

File: app/src/main/java/com/hover/stax/bounties/BountyEmailFragment.kt
package com.hover.stax.bounties

import android.os.Bundle
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.hover.stax.R
import com.hover.stax.databinding.FragmentBountyEmailBinding
import com.hover.stax.home.MainActivity
import com.hover.stax.login.LoginViewModel
import com.hover.stax.settings.SettingsFragment
import com.hover.stax.user.StaxUser
import com.hover.stax.utils.AnalyticsUtil.logAnalyticsEvent
import com.hover.stax.utils.NavUtil
import com.hover.stax.utils.network.NetworkMonitor
import com.hover.stax.views.StaxDialog
import com.uxcam.UXCam
import org.koin.androidx.viewmodel.ext.android.sharedViewModel

class BountyEmailFragment : Fragment(), View.OnClickListener {

    private var _binding: FragmentBountyEmailBinding? = null
    private val binding get() = _binding!!
    private var dialog: StaxDialog? = null
    private lateinit var networkMonitor: NetworkMonitor
    private val loginViewModel: LoginViewModel by sharedViewModel()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentBountyEmailBinding.inflate(inflater, container, false)
        networkMonitor = NetworkMonitor(requireContext())
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.progressIndicator.setVisibilityAfterHide(View.GONE)
        binding.instructions.movementMethod = LinkMovementMethod.getInstance()
        startObservers()
    }

    private fun startObservers() {
        with(loginViewModel) {
            progress.observe(viewLifecycleOwner) { updateProgress(it) }
            error.observe(viewLifecycleOwner) { it?.let { showError(it) } }
            staxUser.observe(viewLifecycleOwner) { initUI(it) }
        }
    }

    private fun initUI(staxUser: StaxUser?) = with(binding) {
        if (staxUser != null && !staxUser.isMapper) {
            btnSignIn.visibility = View.GONE
            joinMappers.apply {
                visibility = View.VISIBLE
                setOnClickListener(this@BountyEmailFragment)
            }
        } else {
            joinMappers.visibility = View.GONE
            btnSignIn.apply {
                visibility = View.VISIBLE
                setOnClickListener(this@BountyEmailFragment)
            }
        }
    }

    override fun onClick(v: View) {
        if (networkMonitor.isNetworkConnected) {
            when (v.id) {
                R.id.btnSignIn -> startGoogleSignIn()
                R.id.joinMappers -> joinMappers()
            }
        } else {
            showDialog(R.string.internet_required, getString(R.string.internet_required_bounty_desc), R.string.btn_ok)
        }
    }

    private fun startGoogleSignIn() {
        logAnalyticsEvent(getString(R.string.clicked_bounty_email_continue_btn), requireContext())
        updateProgress(0)
        (activity as MainActivity).signIn()
        loginViewModel.postGoogleAuthNav.value = SettingsFragment.SHOW_BOUNTY_LIST
    }

    private fun joinMappers() {
        updateProgress(0)
        loginViewModel.joinMappers()
    }

    private fun updateProgress(progress: Int) = with(binding.progressIndicator) {
        when (progress) {
            0 -> show()
            -1 -> hide()
            100 -> {
                hide()
                complete()
            }
            else -> setProgressCompat(progress, true)
        }
    }

    private fun complete() = NavUtil.navigate(findNavController(), BountyEmailFragmentDirections.actionBountyEmailFragmentToBountyListFragment())

    private fun showError(message: String) {
        updateProgress(-1)
        showDialog(0, message, R.string.btn_ok)
    }

    private fun showDialog(title: Int, msg: String, btn: Int) {
        dialog = StaxDialog(requireActivity())
            .setDialogMessage(msg)
            .setPosButton(btn, null)
            .makeSticky()

        if (title != 0)
            dialog?.setDialogTitle(title)

        dialog!!.showIt()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        if (dialog != null && dialog!!.isShowing) dialog!!.dismiss()
        _binding = null
    }

}